src/main/scala/zio/prelude/NonEmptySet.scala

```diff
@@ -0,0 +1,207 @@
+   */
+  def toNonEmptyList: NonEmptyList[A] = destruct match { case (head, tail) => NonEmptyList.fromIterable(head, tail) }
+
+  /** Creates a new `NonEmptySet` with an additional element, unless the element is
+   *  already present.
+   *
+   *  @param elem the element to be added
+   *  @return a new set that contains all elements of this set and that also
+   *          contains `elem`.
+   */
+  def +(elem: A): NonEmptySet[A] = new NonEmptySet(toSet + elem)
+
+  /** Creates a new `NonEmptySet` with additional elements, omitting duplicates.
+   *
+   *  This method takes two or more elements to be added. Elements that already exist in the `NonEmptySet` will
+   *  not be added. Another overloaded variant of this method handles the case where a single element is added.
+   *
+   *  Example:
+   *   {{{
+   *    scala> val a = NonEmptySet(1, 3) + 2 + 3
+   *    a: zio.prelude.NonEmptySet[Int] = NonEmptySet(1, 3, 2)
+   *   }}}
+   *
+   *  @param elem1 the first element to add.
+   *  @param elem2 the second element to add.
+   *  @param elems the remaining elements to add.
+   *  @return   a new `NonEmptySet` with the given elements added, omitting duplicates.
+   */
+  def +(elem1: A, elem2: A, elems: A*): NonEmptySet[A] = new NonEmptySet(toSet + elem1 + elem2 ++ elems)
+
+  /** Computes the union between of `NonEmptySet` and another set.
+   *
+   *  @param   that  the set to form the union with.
+   *  @return  a new `NonEmptySet` consisting of all elements that are in this
+   *  set or in the given set `that`.
+   */
+  def union(that: Set[A]): NonEmptySet[A] = new NonEmptySet(toSet.union(that))
+
+  /** Creates a new `NonEmptySet` by adding all elements contained in another collection to this `NonEmptySet`, omitting duplicates.
+   *
+   * This method takes a collection of elements and adds all elements, omitting duplicates, into `NonEmptySet`.
+   *
+   * Example:
+   *  {{{
+   *    scala> val a = NonEmptySet(1, 2) ++ NonEmptySet(2, "a")
+   *    a: zio.prelude.NonEmptySet[Any] = NonEmptySet(1, 2, a)
+   *  }}}
+   *
+   *  @param elems     the collection containing the elements to add.
+   *  @return a new `NonEmptySet` with the given elements added, omitting duplicates.
+   */
+  def ++(elems: Iterable[A]): NonEmptySet[A] = new NonEmptySet(toSet ++ elems)
+
+  /** Adds the `elem` to this `NonEmptySet`. Alias for `+`. */
+  def add(elem: A): NonEmptySet[A] = self + elem
+
+  /** Removes the `elem` from this `NonEmptySet`. Alias for `-`. */
+  def remove(elem: A): Set[A] = toSet - elem
+
+  /**
+   * Flattens a `NonEmptySet` of `NonEmptySet` values into a single
+   * `NonEmptySet`.
+   */
+  def flatten[B](implicit ev: A <:< NonEmptySet[B]): NonEmptySet[B] =
+    destruct match {
+      case (head, tail) =>
+        tail.foldLeft(ev(head)) { (b, a) =>
+          b ++ ev(a)
+        }
+    }
+
+  override def hashCode: Int = toSet.hashCode()
+
+  override def equals(that: Any): Boolean = that match {
```

Comment:
I think there is a slight inconsistency between the definitions of `hashCode` and `equals` here because a `NonEmptySet` has the same hash code as its underlying set representation but `equals` is saying that a nonempty set is only equal to another nonempty set.

Reply:
That should be fine, shouldn't it? The law that requires equal instances to return equal hashcodes still holds.

Reply:
Yes if values are equal they have the same hash code but a lot of values with the same hash code are not equal which isn't a law violation but is less than ideal. Why don't we add a salt to the hash specific to NonEmptySet so the hash codes of NonEmptySet values are distinct from the hash codes of their underlying representations?

Reply:
I'd be weary of making the `hashCode` method (or `equals`, for that matter) any more complicated than necessary. The more code there is, the more places for a bug to hide.
I like that now it's quite foolproof.

What would be the practical concern, when a `NonEmptySet` and a `Set` with the same content collide on hashCode? In the vast majority of cases, they won't even be in the same hash-based collection (like Map or Set), because they have different type.

Are you sure we want to do that?

Reply:
I think in general wrapper types have different hash codes than the values they wrap. Like `Option(1)` has a different hash code than `1`, `Right(1)`, or `Left(1)`. Most of the time we just get this for free from using case classes because this is how hash code for case classes work but since we aren't using a case class to hide the implementation we have to do it ourselves. I think if we just pick one arbitrary hash, maybe the hash code of "NonEmptySet" and combine it with the hash code of the set itself we should generate different hash codes with minimal work. This is basically what case classes do.

Reply:
fair enough